From c67ec5cd33e968949ce900941e25d2942196f528 Mon Sep 17 00:00:00 2001
From: Patrick Delaunay <patrick.delaunay@foss.st.com>
Date: Mon, 13 Feb 2023 17:46:21 +0100
Subject: [PATCH 438/529] drivers: watchdog: arm_smc_wdt: set WDOG_HW_RUNNING
 at probe

If the watchdog hardware is already enabled during the boot process,
when the Linux watchdog driver loads, it should start/reset the watchdog
and tell the watchdog framework. As a result, ping can be generated from
the watchdog framework (if CONFIG_WATCHDOG_HANDLE_BOOT_ENABLED is set),
until the userspace watchdog daemon takes over control

Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: I4ffc81cd57efe5cb62c24f5a1eb63b3c9e6e3172
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/290809
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Antonio Maria BORNEO <antonio.borneo@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-by: Lionel DEBIEVE <lionel.debieve@foss.st.com>
Tested-by: Christophe PRIOUZEAU <christophe.priouzeau@st.com>
Domain-Review: Lionel DEBIEVE <lionel.debieve@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/299845
Tested-by: Eric FOURMONT <eric.fourmont-ext@st.com>
Reviewed-by: Eric FOURMONT <eric.fourmont-ext@st.com>
Domain-Review: Eric FOURMONT <eric.fourmont-ext@st.com>
---
 drivers/watchdog/arm_smc_wdt.c | 9 +++++++++
 1 file changed, 9 insertions(+)

--- a/drivers/watchdog/arm_smc_wdt.c
+++ b/drivers/watchdog/arm_smc_wdt.c
@@ -148,6 +148,15 @@ static int smcwd_probe(struct platform_d
 	if (err)
 		return err;
 
+	if (IS_ENABLED(CONFIG_WATCHDOG_HANDLE_BOOT_ENABLED)) {
+		err = smcwd_start(wdd);
+		if (err)
+			return err;
+
+		/* Make sure the watchdog is serviced */
+		set_bit(WDOG_HW_RUNNING, &wdd->status);
+	}
+
 	err = devm_watchdog_register_device(&pdev->dev, wdd);
 	if (err)
 		return err;
